File: token-validation-core/src/test/java/no/nav/security/token/support/core/configuration/IssuerConfigurationTest.java
package no.nav.security.token.support.core.configuration;

import com.nimbusds.oauth2.sdk.as.AuthorizationServerMetadata;
import no.nav.security.token.support.core.IssuerMockWebServer;
import no.nav.security.token.support.core.exceptions.MetaDataNotAvailableException;
import no.nav.security.token.support.core.validation.DefaultConfigurableJwtValidator;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.net.URI;
import java.util.List;

import static java.util.Collections.emptyList;
import static no.nav.security.token.support.core.JwtTokenConstants.AUTHORIZATION_HEADER;
import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatExceptionOfType;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

class IssuerConfigurationTest {

    private IssuerMockWebServer issuerMockWebServer;

    @BeforeEach
    void setup() throws IOException {
        issuerMockWebServer = new IssuerMockWebServer(false);
        issuerMockWebServer.start();
    }

    @AfterEach
    void after() throws IOException {
        issuerMockWebServer.shutdown();
    }

    @Test
    void issuerConfigurationWithMetadataFromDiscoveryUrl() {
        IssuerConfiguration config = new IssuerConfiguration(
            "issuer1", new IssuerProperties(issuerMockWebServer.getDiscoveryUrl(), List.of("audience1")), new ProxyAwareResourceRetriever());
        assertThat(config.getMetadata()).isNotNull();
        assertThat(config.getTokenValidator()).isNotNull();
        assertThat(config.getTokenValidator()).isInstanceOf(DefaultConfigurableJwtValidator.class);
        AuthorizationServerMetadata metadata = config.getMetadata();
        assertThat(metadata.getIssuer()).isNotNull();
        assertThat(metadata.getJWKSetURI().toString()).isNotNull();
    }

    @Test
    void issuerConfigurationDiscoveryUrlNotValid() {
        assertThatExceptionOfType(MetaDataNotAvailableException.class).isThrownBy(() -> new IssuerConfiguration(
            "issuer1",
            new IssuerProperties(URI.create("http://notvalid").toURL(), List.of("audience1")),
            new ProxyAwareResourceRetriever()));
        assertThatExceptionOfType(MetaDataNotAvailableException.class).isThrownBy(() -> new IssuerConfiguration(
            "issuer1",
            new IssuerProperties(URI.create("http://localhost").toURL(), List.of("audience1")),
            new ProxyAwareResourceRetriever()));
        assertThatExceptionOfType(MetaDataNotAvailableException.class).isThrownBy(() -> new IssuerConfiguration(
            "issuer1",
            new IssuerProperties(URI.create(issuerMockWebServer.getDiscoveryUrl().toString() + "/pathincorrect").toURL(),
                List.of("audience1")),
            new ProxyAwareResourceRetriever()));
    }

    @Test
    void issuerConfigurationWithConfigurableJwtTokenValidator() {
        IssuerProperties issuerProperties = new IssuerProperties(
            issuerMockWebServer.getDiscoveryUrl(), emptyList(),null,AUTHORIZATION_HEADER, new IssuerProperties.Validation(List.of("sub", "aud"))
        );
        IssuerConfiguration config = new IssuerConfiguration(
            "issuer1",
            issuerProperties,
            new ProxyAwareResourceRetriever()
        );
        assertThat(config.getMetadata()).isNotNull();
        assertThat(config.getTokenValidator()).isNotNull();
        assertThat(config.getTokenValidator()).isInstanceOf(DefaultConfigurableJwtValidator.class);
        AuthorizationServerMetadata metadata = config.getMetadata();
        assertThat(metadata.getIssuer()).isNotNull();
        assertThat(metadata.getJWKSetURI().toString()).isNotNull();
        assertTrue(issuerProperties.getValidation().isConfigured());
    }

    @Test
    void issuerConfigurationWithConfigurableJWKSCacheAndConfigurableJwtTokenValidator() {
        IssuerProperties issuerProperties = new IssuerProperties(
            issuerMockWebServer.getDiscoveryUrl(),emptyList(),null,AUTHORIZATION_HEADER,
            new IssuerProperties.Validation(List.of("sub", "aud")),
            new IssuerProperties.JwksCache(15L, 5L)
        );
        IssuerConfiguration config = new IssuerConfiguration(
            "issuer1",
            issuerProperties,
            new ProxyAwareResourceRetriever()
        );
        assertThat(config.getMetadata()).isNotNull();
        assertThat(config.getTokenValidator()).isNotNull();
        assertThat(config.getTokenValidator()).isInstanceOf(DefaultConfigurableJwtValidator.class);
        AuthorizationServerMetadata metadata = config.getMetadata();
        assertThat(metadata.getIssuer()).isNotNull();
        assertThat(metadata.getJWKSetURI().toString()).isNotNull();
        assertTrue(issuerProperties.getJwksCache().isConfigured());
        assertTrue(issuerProperties.getValidation().isConfigured());
    }
}

File: token-validation-core/src/test/kotlin/no/nav/security/token/support/core/configuration/IssuerConfigurationTest.kt
package no.nav.security.token.support.core.configuration

import com.nimbusds.jwt.JWTClaimNames.AUDIENCE
import com.nimbusds.jwt.JWTClaimNames.SUBJECT
import java.net.URI
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import no.nav.security.token.support.core.IssuerMockWebServer
import no.nav.security.token.support.core.JwtTokenConstants.AUTHORIZATION_HEADER
import no.nav.security.token.support.core.configuration.IssuerProperties.JwksCache
import no.nav.security.token.support.core.configuration.IssuerProperties.Validation
import no.nav.security.token.support.core.exceptions.MetaDataNotAvailableException

internal class IssuerConfigurationTest {

    private lateinit var issuerMockWebServer : IssuerMockWebServer
    @BeforeEach
    fun setup() {
        issuerMockWebServer = IssuerMockWebServer(false).apply {
            start()
        }
    }

    @AfterEach
    fun after() {
        issuerMockWebServer.shutdown()
    }

    @Test
    fun issuerConfigurationWithMetadataFromDiscoveryUrl() {
        assertThat(IssuerConfiguration("issuer1", IssuerProperties(issuerMockWebServer.discoveryUrl, listOf("audience1")),
            ProxyAwareResourceRetriever()).metadata)
            .extracting(
                { it.issuer != null },
                { it.jwkSetURI != null })
            .containsExactly(true, true)
    }

    @Test
    fun issuerConfigurationDiscoveryUrlNotValid() {
        assertThrows<MetaDataNotAvailableException> {
            IssuerConfiguration("issuer1", IssuerProperties(URI("http://notvalid").toURL(), listOf("audience1")))
        }
        assertThrows<MetaDataNotAvailableException> {
            IssuerConfiguration("issuer1", IssuerProperties(URI("http://localhost").toURL(), listOf("audience1")))
        }
        assertThrows<MetaDataNotAvailableException> {
            IssuerConfiguration("issuer1", IssuerProperties(URI.create(issuerMockWebServer.discoveryUrl.toString() + "/pathincorrect").toURL(), listOf("audience1")), ProxyAwareResourceRetriever())
        }
    }

    @Test
    fun issuerConfigurationWithConfigurableJwtTokenValidator() {
        val p = IssuerProperties(issuerMockWebServer.discoveryUrl, emptyList(), null, AUTHORIZATION_HEADER, Validation(listOf(SUBJECT, AUDIENCE)))
        assertThat(IssuerConfiguration("issuer1", p).metadata)
            .extracting(
                { it.issuer != null },
                { it.jwkSetURI != null })
            .containsExactly(true, true)
        assertTrue(p.validation.isConfigured)
    }

    @Test
    fun issuerConfigurationWithConfigurableJWKSCacheAndConfigurableJwtTokenValidator() {
        val p = IssuerProperties(issuerMockWebServer.discoveryUrl, emptyList(), null, AUTHORIZATION_HEADER, Validation(listOf(SUBJECT, AUDIENCE)), JwksCache(15L, 5L))
        assertThat(IssuerConfiguration("issuer1", p).metadata)
            .extracting(
                { it.issuer != null },
                { it.jwkSetURI != null })
            .containsExactly(true, true)
        assertThat(p)
            .extracting(
                { it.jwksCache.isConfigured},
                { it.validation.isConfigured })
            .containsExactly(true, true)
    }
}